### PR TITLE
[Constraint System] Implement heuristics for linked operator expressions in the solver proper.

### PR DESCRIPTION
**Pull request by hborla**
This PR contains the first handful of commits from @gregomni in https://github.com/apple/swift/pull/26140, with some adjustments. Specifically:
* If an operator overload choice has the same type as an existing overload binding elsewhere in the expression, attempt that choice first. This is done by ordering such choices first in the main disjunction partition rather than using the favoring mechanism, because the solver will stop if it successfully finds a solution with a favored constraint, which could be incorrect.
* Consider the number of favored choices when choosing which disjunction to try first in `selectDisjunction`.

These heuristics together help the solver reach a solution fast for expressions with multiple operators chained together where the operands all have the same type. 

Finding a solution fast allows the solver to do more aggressive pruning later on. Once this change lands, I'm going to work on pruning heuristics for generic overloads, which I think will finally let us remove `LinkedExprAnalyzer` and the hacks around favored constraints. Once those hacks are removed, the favoring mechanism can be used to suggest choices that are likely to work without worrying about correctness.

All thanks to Greg for the approach taken in this PR!!

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test compiler performance

**Comment by hborla** [comment]
@swift-ci please test compiler performance

**Comment by hborla** [comment]
@swift-ci please smoke test compiler performance

**Comment by hborla** [comment]
@swift-ci please test compiler performance

**Comment by hborla** [comment]
@swift-ci please test compiler performance

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
@swift-ci please test

**Comment by hborla** [comment]
@swift-ci please test

**Comment by hborla** [comment]
@swift-ci please test compiler performance

**Review by xedin — COMMENTED** [review]
Makes sense to me! I have left a couple of comments inline and I wonder whether performance regressions are related to the fact that solver would pull generic overloads up before attempting the ones with concrete types which means that it ends up exploring a lot more space than before. It'd be great to get them added to the suite.

File: lib/Sema/CSSolver.cpp [hunk 1]
```diff
@@ -2013,6 +2013,51 @@ static Constraint *tryOptimizeGenericDisjunction(
...
+    typesFound.insert(representativeDecl->getInterfaceType()->getCanonicalType());
+  }
+
+  for (auto index : indices(constraints)) {
+    auto *constraint = constraints[index];
+    if (constraint->isFavored())
+      continue;
+
+    auto *decl = constraint->getOverloadChoice().getDecl();
+    if (typesFound.count(decl->getInterfaceType()->getCanonicalType()))
```
[the remark this answers is not in the document]
Comment (hborla) [reply lib/Sema/CSSolver.cpp:2056]: Ah, that's a good point that I hadn't thought of. I do think it only makes sense to do this if the requirements are the same...or maybe we should just restrict this heuristic to the exact same decl in the case of generics. I think the "same type" piece is most useful for concrete overloads

**Comment by xwu** [comment]
@hborla, I just want to put on your radar the following issue if you weren't already aware of it (I forget which SR was filed about it, but I think there is one). It would have been fixed by "designated operators," which is why I'm rather sad it didn't make it:

SE-0104 added heterogeneous comparison and bit shift operators for integer types in order to improve the user experience. These heterogeneous operators behave as intended with concrete types. However, integer comparisons in generic algorithms behave differently (and were source breaking too):

If one operand in a generic context is a literal, overload resolution will favor heterogeneous comparison with `IntegerLiteralType` over homogeneous comparison.

The same problem would occur in a nongeneric context but for two hacks: (1) to improve type-checking performance, the compiler won’t traverse the protocol hierarchy to rank all overloads of an operator function if it finds a matching overload defined in the concrete type; (2) to favor homogeneous comparison over heterogeneous comparison, [otherwise redundant concrete implementations](https://github.com/apple/swift/pull/9909) have been added to built-in numeric types.

This issue was encountered during review of the standard library’s implementation of `DoubleWidth`, which in fact [had a bug](https://github.com/apple/swift/pull/9367#discussion_r118612475) as a consequence of the following behavior:

```swift
func f() -> Bool {
  return UInt.max == ~0
}
func g<T : FixedWidthInteger>(_: T.Type) -> Bool {
  return T.max == ~0
}

f()          // `true`
g(UInt.self) // `false`
```

In a generic context, even if the most refined protocol implements its own overload of the homogeneous comparison operator, Swift will look for all overloads of the operator by traversing the entire protocol hierarchy. Since heterogeneous comparison operators are defined somewhere along the hierarchy, the compiler will always find an overload that takes the “preferred” `IntegerLiteralType` and therefore infers the literal to be of type `Int`. As a result, in the expression `g(UInt.self)`, we are actually comparing `UInt.max` to `~(0 as Int)`.

Heterogeneous comparison were also planned for floating-point types. However, until the issue described above is resolved, adding heterogeneous comparison would cause new unexpected results:

```swift
func isOnePointTwo<T: BinaryFloatingPoint>(_ value: T) -> Bool {
  return value == 1.2
}

isOnePointTwo(1.2 as Float80)
// Currently `true`.
// Would be `false` when heterogeneous comparison is implemented.
```

Even before the constraints of ABI stability, I had tried a number of solutions at the level of the standard library to get the compiler to prefer homogeneous comparison, but it became apparent that there was no solution without tackling operator expressions in the constraint system.

File: lib/Sema/CSSolver.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/Sema/CSSolver.cpp:2056]: Wouldn't the interface type be written in terms of `GenericFunctionType` in that case, which should include the requirements?

That being said, an identity check on the decl may be cheaper than computing the canonical type for a `GenericFunctionType` – as IIRC that may involve spinning up a new GSB (which was the motivation behind https://github.com/apple/swift/pull/32034).

[the remark this answers is not in the document]
Comment (hborla) [reply lib/Sema/CSSolver.cpp:2056]: > Wouldn't the interface type be written in terms of GenericFunctionType in that case, which should include the requirements?

I know `GenericFunctionType` stores the generic signature which has the requirements, but I'm not sure whether comparing canonical types factors in requirements.

> That being said, an identity check on the decl may be cheaper than computing the canonical type for a GenericFunctionType 

Yeah, I think I'm just going to stick to a `ValueDecl` check in the case of generics. I think it's a lot more common to have mixed operator expressions that use concrete overloads of the same type, so this heuristic for generics is probably more work than it's worth.

**Comment by hborla** [comment]
@swift-ci please smoke test compiler performance

**Comment by hborla** [comment]
@swift-ci please test compiler performance

**Comment by hborla** [comment]
@swift-ci please build toolchain

**Comment by hborla** [comment]
@swift-ci please test source compatibility

**Comment by hborla** [comment]
@swift-ci please smoke test compiler performance

**Comment by hborla** [comment]
@swift-ci please test compiler performance

File: lib/Sema/CSSolver.cpp [hunk 2]
```diff
@@ -2013,6 +2013,60 @@ static Constraint *tryOptimizeGenericDisjunction(
...
+
+  auto overload = choice->getOverloadChoice();
+  if (!overload.isDecl())
+    return;
+  auto decl = overload.getDecl();
+  if (!decl->isOperator())
+    return;
+
+  SmallSet<CanType, 4> concreteTypesFound;
+  SmallSet<ValueDecl *, 4> genericDeclsFound;
```
Comment (hamishknight) [inline lib/Sema/CSSolver.cpp:2038]: Would it be worth adding a comment to explain the difference in handling for generic overloads?

**Review by hamishknight — COMMENTED** [review]
So happy to see more of these solving hacks removed!

File: lib/Sema/CSSolver.cpp (hunk 2 above)
Comment (hborla) [reply lib/Sema/CSSolver.cpp:2044]: Yes, you're right that this needs an explanation. I added a comment - let me know if you think it needs more clarification!

File: validation-test/Sema/type_checker_perf/slow/mixed_string_array_addition.swift [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// RUN: %target-typecheck-verify-swift -swift-version 5 -solver-expression-time-threshold=1
+
+func method(_ arg: String, body: () -> [String]) {}
+
+func test(str: String, properties: [String]) {
+  // expected-error@+1 {{the compiler is unable to type-check this expression in reasonable time}}
```
Comment (hborla) [inline validation-test/Sema/type_checker_perf/slow/mixed_string_array_addition.swift:6]: This isn't new, but a case I discovered when debugging the source compat failures. This compiles (very slowly) if you add a type annotation to `param` below.

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please test compiler performance

File: lib/Sema/CSSolver.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Sema/CSSolver.cpp:2044]: Looks good to me!

**Comment by hborla** [comment]
@swift-ci please smoke test OS X platform

**Comment by hborla** [comment]
@swift-ci please smoke test

**Pull request merged**